File: 1.2/python/main.py

freq = 0
found_freqs = set()

buffer = ''
with open('./../input.txt', 'r') as f:
    buffer = f.read()

found_duplicate = False
iteration_count = 0
while not found_duplicate:
    for line in buffer.split('\n'):
        if line[0] == '+':
            freq += int(line[1:])
        else:
            freq -= int(line[1:])
        if freq in found_freqs:
            found_duplicate = True
            break
        else: 
            found_freqs.add(freq)
    iteration_count += 1
    print('Searching... iteration ' + str(iteration_count))


print('Result frequency: ' + str(freq))

File: 1.2/rust/src/main.rs
use std::fs::File;
use std::io::prelude::*;
use std::string::String;
use std::collections::HashSet;

fn main() {
    let mut freq: i64 = 0;
    let mut found_freqs: HashSet<i64> = HashSet::new();

    let mut raw_frequencies = String::new();
    File::open("./../input.txt")
        .expect("Unable to read from input.txt")
        .read_to_string(&mut raw_frequencies)
        .expect("Unable to feed contents to string");

    let mut iteration_count = 0;
    'repeat: loop {
        println!("Searching... iteration {}", iteration_count);

        'lines: for raw_freq in raw_frequencies.lines() {
            match &raw_freq[..1] {
                "+" => freq += &raw_freq[1..].parse::<i64>().unwrap(),
                "-" => freq -= &raw_freq[1..].parse::<i64>().unwrap(),
                _ => println!("Parsed unexpected value... ignoring"),
            };

            if found_freqs.contains(&freq) {
                println!("Found repeating frequency");
                break 'repeat;
            }

            &found_freqs.insert(freq);
        }

        iteration_count += 1;
    }
    println!("Resulting Frequency: {}", freq);
}
